docs/docs/building-modules/09-autocli.md

```diff
@@ -0,0 +1,174 @@
+
+`autocli` generates flags for each field in a protobuf message. By default, the names of the flags are generated based on the names of the fields in the message. You can customise the flag names using the `namingOptions` parameter of the `Builder.AddMessageFlags()` method.
+
+To define flags for a message, you can use the `Builder.AddMessageFlags()` method. This method takes the `cobra.Command` instance and the message type as input, and generates flags for each field in the message.
+
+``` go reference
+binder, err := b.AddMessageFlags(cmd.Context(), cmd.Flags(), inputType, options)
+if err != nil {
+    return nil, err
+}
+
+cmd.Args = binder.CobraArgs
+
+```
+
+The `binder` variable returned by the `AddMessageFlags()` method is used to bind the command-line arguments to the fields in the message.
+
+You can also customise the behavior of the flags using the `namingOptions` parameter of the `Builder.AddMessageFlags()` method. This parameter allows you to specify a custom prefix for the flags, and to specify whether to generate flags for repeated fields and whether to generate flags for fields with default values.
+
+## Commands and Queries
+
+The `autocli` package generates CLI commands and flags for each method defined in your gRPC service. By default, it generates commands for each RPC method that does not return a stream of messages. The commands are named based on the name of the service method.
+
+For example, given the following protobuf definition for a service:
+
+```protobuf
+service MyService {
+  rpc MyMethod(MyRequest) returns (MyResponse) {}
+}
+
+```
+
+`autocli` will generate a command named `my-method` for the `MyMethod` method. The command will have flags for each field in the `MyRequest` message.
+
+If you want to customise the behavior of a command, you can define a custom command by implementing the `autocli.Command` interface. You can then register the command with the `autocli.Builder` instance for your application.
+
+Similarly, you can define a custom query by implementing the `autocli.Query` interface. You can then register the query with the `autocli.Builder` instance for your application.
+
+To add a custom command or query, you can use the `Builder.AddCustomCommand` or `Builder.AddCustomQuery` methods, respectively. These methods take a `cobra.Command` or `cobra.Command` instance, respectively, which can be used to define the behavior of the command or query.
+
+## Advanced Usage
+
+### Specifying Subcommands
+
+By default, `autocli` generates a command for each method in your gRPC service. However, you can specify subcommands to group related commands together. To specify subcommands, you can use the `autocliv1.ServiceCommandDescriptor` struct.
+
+For example, suppose you have a service with two methods: `GetTest` and `GetTest2`. You can group these methods together under a `get` subcommand using the following code
+
+```go reference
+ // AutoCLIOptions implements the autocli.HasAutoCLIConfig interface. 
```

Comment:
For references, we actually need a GitHub link, with the lines containing the snippet :)
This won't work.

Reply:
lol haha will add now